src/domain/RBF/rbf.cpp

```diff
@@ -0,0 +1,742 @@
+    MatSetOption(A, MAT_SYMMETRIC, PETSC_TRUE) >> utilities::PetscUtilities::checkError;
+
+    MatDenseGetArrayWrite(A, &vals) >> utilities::PetscUtilities::checkError;
+
+    // RBF contributions to the matrix
+    for (i = 0; i < nCells; ++i) {
+        for (j = i; j < nCells; ++j) {
+            vals[i * matSize + j] = vals[j * matSize + i] = RBFVal(dim, &x[i * dim], &x[j * dim]);
+        }
+    }
+
+    if (nPoly > 0) {
+        // Augmented polynomial contributions
+        switch (dim) {
+            case 1:
+                for (i = 0; i < nCells; ++i) {
+                    j = nCells;
+                    for (px = 0; px < p1; ++px) {
+                        vals[i * matSize + j] = vals[j * matSize + i] = xp[(i * dim + 0) * p1 + px];
+                        ++j;
+                    }
+                }
+                break;
+            case 2:
+                for (i = 0; i < nCells; ++i) {
+                    j = nCells;
+                    for (py = 0; py < p1; ++py) {
+                        for (px = 0; px < p1 - py; ++px) {
+                            vals[i * matSize + j] = vals[j * matSize + i] = xp[(i * dim + 0) * p1 + px] * xp[(i * dim + 1) * p1 + py];
+                            ++j;
+                        }
+                    }
+                }
+                break;
+            case 3:
+                for (i = 0; i < nCells; ++i) {
+                    j = nCells;
+                    for (pz = 0; pz < p1; ++pz) {
+                        for (py = 0; py < p1 - pz; ++py) {
+                            for (px = 0; px < p1 - py - pz; ++px) {
+                                vals[i * matSize + j] = vals[j * matSize + i] = xp[(i * dim + 0) * p1 + px] * xp[(i * dim + 1) * p1 + py] * xp[(i * dim + 2) * p1 + pz];
+                                ++j;
+                            }
+                        }
+                    }
+                }
+                break;
+            default:
+                throw std::runtime_error("ablate::domain::RBF::Matrix encountered an unknown dimension.");
+        }
+    }
+
+    MatDenseRestoreArrayWrite(A, &vals) >> utilities::PetscUtilities::checkError;
+    MatViewFromOptions(A, NULL, "-ablate::domain::rbf::RBF::A_view") >> utilities::PetscUtilities::checkError;
+
+    // Factor the matrix
+    MatLUFactor(A, NULL, NULL, NULL) >> utilities::PetscUtilities::checkError;
+
+    PetscFree(xp) >> utilities::PetscUtilities::checkError;
+
+    // Assign output
+    RBF::RBFMatrix[c] = A;
+    RBF::stencilXLocs[c] = x;
+}
+
+/************ Begin Derivative Code **********************/
+
+// Set the derivatives to use
+// nDer - Number of derivatives to set
+// dx, dy, dz - Lists of length nDer indicating the derivatives
+// useVertices - Use common vertices when determining neighbors. If false then use common edges.
+void RBF::SetDerivatives(PetscInt nDer, PetscInt dx[], PetscInt dy[], PetscInt dz[], PetscBool useVertices) {
```

Comment:
Would you mind changing the local variables nPoly, nStencil, nDer and dxyz?  I'm getting a lot of "Declaration shadows a field of 'ablate::domain::rbf::RBF'" warnings.